### PR TITLE
Linux/ARM: Use -O1 level to avoid segfault in release-build

### PR DESCRIPTION
**Pull request by leemgs**
We cannot still run 'hello world' console application even though
the cross compilation of CoreCLR is successfully completed.
In release-build, the segmentation fault of 'hello world'
is made by the misalignment of thread local storage (TLS) section
because of the aggressive optimization level of clang compiler for
code optimization(e.g., file size and execution speed).
It means that Clang/LLVM has a bug in case of the usage of O2/O3 flag.

Below is the major difference among the optimization levels of Clang.
* -O2 is based on -O1:
  .adding: -gvn -constmerge -globaldce -slp-vectorizer -mldst-motion -inline
  .removing: -always-inline
* -O3 is based on -O2:
  .adding: -argpromotion

ver2:
- Corerun is loading 'libcoreclr.so' using dlopen() library call. So, we can not
  use initial-exec TLS model (ver1) because of the thread-safe issue of
  the multi-threaded. (Reported by @janvorli)
- In the release-build, Let's replace -O3 with -O1 to avoid the segmentation
  fault due to the aggressive optimization level of the Clang until fixing
  the bug of the Clang/LLVM.

ver1:
- The default value of clang is "global-dynamic". However, the thread-local
storage (TLS) model of Clang/LLVM does not guarantee the normal execution
of TLS's symbol relocation due to the misaligned __tls_get_addr symbol in
case of the aggressive optimizations of Clang on Linux/ARM.
- Let's enable initial-exec TLS model instead of the dynamic TLS model.

Signed-off-by: Geunsik Lim <geunsik.lim@samsung.com>
CC: Jan Kotas <jkotas@microsoft.com>
CC: Jan Vorlicek <janvorli@microsoft.com>

### CODE REVIEW

**Comment by leemgs** [comment]
This PR is related to https://github.com/dotnet/coreclr/pull/4715.

Below is experimental result. 

* Before this PR:
```
runtime-arm-u1404-release-mode-20160511# time ./corerun /dotnet/example/hello.exe
Segmentation fault

real	0m1.469s
user	0m0.160s
sys	0m0.140s
```

* After this PR:
  - **5.987s** in release-build, **24.448s** in debug-build. 
```
runtime-arm-u1404-release-mode-20160511# time ./corerun /dotnet/example/hello.exe
Hello World

real	0m5.987s
user	0m5.330s
sys	0m0.350s
```



\CC: @jkotas , @janvorli 

> git blame  ./src/pal/tools/clang-compiler-override.txt  | grep RELEASE
> 1e86aaad (Mike McLaughlin  2016-03-01 09:26:07 -0800  4) SET (CMAKE_C_FLAGS_RELEASE_INIT        "-g -O3")
> 1e86aaad (Mike McLaughlin  2016-03-01 09:26:07 -0800 10) SET (CMAKE_CXX_FLAGS_RELEASE_INIT        "-g -O3")
> ^ef1e2ab (dotnet-bot       2015-01-30 14:14:42 -0800 15) SET (CLR_DEFINES_RELEASE_INIT            NDEBUG URTBLDENV_FRIENDLY=Retail)

\CC: @mikem8361

File: src/pal/tools/clang-compiler-override.txt [hunk 1]
```diff
@@ -1,13 +1,23 @@
 SET (CMAKE_C_FLAGS_INIT                "-Wall -std=c11")
 SET (CMAKE_C_FLAGS_DEBUG_INIT          "-g -O0")
 SET (CLR_C_FLAGS_CHECKED_INIT          "-g -O2")
-SET (CMAKE_C_FLAGS_RELEASE_INIT        "-g -O3")
+# Todo: Replace -O1 with -O3 after fixing Clang/LLVM for __thread on Linux/ARM
+if(CMAKE_SYSTEM_NAME STREQUAL Linux AND CMAKE_SYSTEM_PROCESSOR STREQUAL armv7l)
+    SET (CMAKE_C_FLAGS_RELEASE_INIT        "-g -O1")
+else()
+    SET (CMAKE_C_FLAGS_RELEASE_INIT        "-g -O3")
+    message(FATAL_ERROR "RELEASE... c flag no-arm")
```
Comment (jkotas) [inline src/pal/tools/clang-compiler-override.txt]: Is this a left over logging?

Comment (leemgs) [inline src/pal/tools/clang-compiler-override.txt]: Fixed. Thanks.

**Comment by janvorli** [comment]
LGTM

**Pull request merged**